src/audio/crossover/crossover.c

```diff
@@ -0,0 +1,745 @@
+
+	/* Assign LR4 coefficients from config */
+	ret = crossover_init_coef(cd, nch);
+
+	return ret;
+}
+
+/**
+ * \brief Creates a Crossover Filter component.
+ * \return Pointer to Crossover Filter component device.
+ */
+static struct comp_dev *crossover_new(const struct comp_driver *drv,
+				      struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct comp_data *cd;
+	struct sof_ipc_comp_process *crossover;
+	struct sof_ipc_comp_process *ipc_crossover =
+			(struct sof_ipc_comp_process *)comp;
+	size_t bs = ipc_crossover->size;
+	int ret;
+
+	comp_cl_info(&comp_crossover, "crossover_new()");
+
+	/* Check that the coefficients blob size is sane.
+	 * It should have the coefficients for 6 biquads.
+	 * TODO Have the check somehow
+	 */
+	if (bs > 0) {
+		comp_cl_err(&comp_crossover, "crossover_new(), NOT IMPLEMENTED YET");
+		return NULL;
+	}
+
+	dev = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+		      COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+	dev->drv = drv;
+
+	dev->size = COMP_SIZE(struct sof_ipc_comp_process);
+
+	crossover = COMP_GET_IPC(dev, sof_ipc_comp_process);
+	ret = memcpy_s(crossover, sizeof(*crossover), ipc_crossover,
+		       sizeof(struct sof_ipc_comp_process));
+	assert(!ret);
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0,
+		     SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	cd->crossover_func = NULL;
+	cd->config = NULL;
+	cd->config_new = NULL;
+
+	if (bs) {
+		cd->config = rzalloc(SOF_MEM_ZONE_RUNTIME, 0,
+				     SOF_MEM_CAPS_RAM, bs);
+		if (!cd->config) {
+			rfree(dev);
+			rfree(cd);
+			return NULL;
+		}
+
+		ret = memcpy_s(cd->config, bs, ipc_crossover->data, bs);
+		assert(!ret);
+	}
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+/**
+ * \brief Frees Crossover Filter component.
+ * \param[in,out] dev Crossover filter base component device.
+ */
+static void crossover_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_info(dev, "crossover_free()");
+
+	crossover_free_config(&cd->config);
+	crossover_free_config(&cd->config_new);
+
+	crossover_reset_state(cd);
+
+	rfree(cd);
+	rfree(dev);
+}
+
+/**
+ * \brief Sets Crossover Filter component audio stream parameters.
+ * \param[in,out] dev Crossover Filter base component device.
+ * \return Error code.
+ *
+ * All done in prepare() since we need to know source and sink component params.
+ */
+static int crossover_params(struct comp_dev *dev,
+			    struct sof_ipc_stream_params *params)
+{
+	comp_info(dev, "crossover_params()");
+
+	return 0;
+}
+
+static int crossover_cmd_get_data(struct comp_dev *dev,
+				  struct sof_ipc_ctrl_data *cdata, int max_size)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	size_t bs;
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "crossover_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+
+		/* Copy back to user space */
+		if (cd->config) {
```

Comment:
I'm seriously thinking there should be some helpers written for this boiler plate code. Hmm maybe I will write a PR, just a personal thought.